File: Infer.ts
import { Expression, Op } from "./Parser.ts";
import { solver } from "./Solver.ts";
import {
  createFresh,
  Scheme,
  TArr,
  Type,
  typeBool,
  TypeEnv,
  typeError,
  typeInt,
} from "./Typing.ts";

export class Constraints {
  constraints: Array<[Type, Type]> = [];

  add(t1: Type, t2: Type) {
    this.constraints.push([t1, t2]);
  }
}

const ops = new Map([
  [Op.Equals, new TArr(typeInt, new TArr(typeInt, typeBool))],
  [Op.Plus, new TArr(typeInt, new TArr(typeInt, typeInt))],
  [Op.Minus, new TArr(typeInt, new TArr(typeInt, typeInt))],
  [Op.Times, new TArr(typeInt, new TArr(typeInt, typeInt))],
  [Op.Divide, new TArr(typeInt, new TArr(typeInt, typeInt))],
]);

export const inferExpression = (
  env: TypeEnv,
  expression: Expression,
): [Constraints, Type] => {
  const constraints = new Constraints();
  const pump = createFresh();

  const infer = (env: TypeEnv, expr: Expression): Type => {
    if (expr.type === "App") {
      const t1 = infer(env, expr.e1);
      const t2 = infer(env, expr.e2);
      const tv = pump.next();

      constraints.add(t1, new TArr(t2, tv));

      return tv;
    }
    if (expr.type === "If") {
      const tg = infer(env, expr.guard);
      const tt = infer(env, expr.then);
      const et = infer(env, expr.else);

      constraints.add(tg, typeBool);
      constraints.add(tt, et);

      return tt;
    }
    if (expr.type === "Lam") {
      const tv = pump.next();
      const t = infer(env.extend(expr.name, new Scheme([], tv)), expr.expr);
      return new TArr(tv, t);
    }
    if (expr.type === "Let") {
      const tb = infer(env, expr.body);

      const subst = solver(constraints.constraints);

      const newEnv = env.apply(subst);
      const sc = newEnv.generalise(tb.apply(subst));

      return infer(newEnv.extend(expr.name, sc), expr.expr);
    }
    if (expr.type === "LBool") {
      return typeBool;
    }
    if (expr.type === "LInt") {
      return typeInt;
    }
    if (expr.type === "Op") {
      const tl = infer(env, expr.left);
      const tr = infer(env, expr.right);
      const tv = pump.next();

      const u1 = new TArr(tl, new TArr(tr, tv));
      const u2 = ops.get(expr.op)!;
      constraints.add(u1, u2);
      return tv;
    }
    if (expr.type === "Var") {
      const scheme = env.scheme(expr.name);

      if (scheme === undefined) throw `Unknown name: ${expr.name}`;

      return scheme.instantiate(pump);
    }

    return typeError;
  };

  return [constraints, infer(env, expression)];
};

File: kotlin-stlc/app/src/main/kotlin/stlc/Infer.kt
package stlc

data class InferResult(val constraints: Constraints, val type: Type)

fun infer(typeEnv: TypeEnv, e: Expression): InferResult {
    val state = Infer()

    val type = state.apply(typeEnv, e)

    return InferResult(state.constraints, type)
}

private class Infer(val constraints: Constraints = Constraints(), val pump: Pump = Pump()) {
    fun apply(typeEnv: TypeEnv, e: Expression): Type =
        when (e) {
            is AppExpression -> {
                val t1 = apply(typeEnv, e.e1)
                val t2 = apply(typeEnv, e.e2)
                val tv = pump.next()

                constraints.add(t1, TArr(t2, tv))

                tv
            }

            is IfExpression -> {
                val t1 = apply(typeEnv, e.e1)
                val t2 = apply(typeEnv, e.e2)
                val t3 = apply(typeEnv, e.e3)

                constraints.add(t1, typeBool)
                constraints.add(t2, t3)

                t2
            }

            is LamExpression -> {
                val tv = pump.next()
                val t = apply(typeEnv + Pair(e.name, Scheme(setOf(), tv)), e.e)

                TArr(tv, t)
            }

            is LBoolExpression ->
                typeBool

            is LIntExpression ->
                typeInt

            is LetExpression -> {
                var newTypeEnv = typeEnv

                for (decl in e.declarations) {
                    val tb = apply(newTypeEnv, decl.e)
                    val subst = constraints.solve()
                    newTypeEnv = newTypeEnv.apply(subst)
                    val sc = newTypeEnv.generalise(tb.apply(subst))
                    newTypeEnv = newTypeEnv.extend(decl.name, sc)
                }

                apply(newTypeEnv, e.e)
            }

            is LetRecExpression -> {
                var newTypeEnv = typeEnv
                val declarationBindings = mutableMapOf<String, Type>()

                for (decl in e.declarations) {
                    val type = pump.next()
                    declarationBindings[decl.name] = type
                    newTypeEnv += Pair(decl.name, newTypeEnv.generalise(type))
                }

                for (decl in e.declarations) {
                    val tb = apply(newTypeEnv, decl.e)
                    constraints.add(tb, declarationBindings[decl.name]!!)
                }

                val subst = constraints.solve()
                newTypeEnv = newTypeEnv.apply(subst)

                for (decl in e.declarations) {
                    val sc = newTypeEnv.generalise(declarationBindings[decl.name]!!.apply(subst))
                    newTypeEnv += Pair(decl.name, sc)
                }

                apply(newTypeEnv, e.e)
            }

            is OpExpression -> {
                val t1 = apply(typeEnv, e.left)
                val t2 = apply(typeEnv, e.right)
                val tv = pump.next()

                val u1 = TArr(t1, TArr(t2, tv))
                val u2 = ops[e.op] ?: typeError
                constraints.add(u1, u2)

                tv
            }

            is VarExpression -> {
                val scheme = typeEnv[e.name] ?: throw UnknownNameException(e.name, typeEnv)

                scheme.instantiate(pump)
            }
        }
}

val ops = mapOf<Op, Type>(
    Pair(Op.Equals, TArr(typeInt, TArr(typeInt, typeBool))),
    Pair(Op.Plus, TArr(typeInt, TArr(typeInt, typeInt))),
    Pair(Op.Minus, TArr(typeInt, TArr(typeInt, typeInt))),
    Pair(Op.Times, TArr(typeInt, TArr(typeInt, typeInt))),
    Pair(Op.Divide, TArr(typeInt, TArr(typeInt, typeInt))),
)

data class UnknownNameException(val name: String, val typeEnv: TypeEnv) : Exception()
